krm-functions/lib/fn-runtime/downstream_runtime.go

```diff
@@ -0,0 +1,164 @@
+		}
+
+		// populate condition inventory
+		for _, c := range kf.GetConditions() {
+			// based on the ForObj determine if there is work to be done
+			if strings.Contains(c.Type, kptfilelibv1.GetConditionType(&r.cfg.For.ObjectRef)) {
+				if c.Status == kptv1.ConditionFalse {
+					if len(r.cfg.Owns) == 0 {
+						// this is looking for all resources
+						r.inv.AddForCondition(Wildcard, &c)
+					} else {
+						objRef := kptfilelibv1.GetGVKNFromConditionType(c.Reason)
+						r.inv.AddForCondition(objRef.Name, &c)
+					}
+				}
+			}
+		}
+
+		// collect all conditions
+		for _, c := range kf.GetConditions() {
+			objRef := *kptfilelibv1.GetGVKNFromConditionType(c.Type)
+			if len(r.cfg.Owns) == 0 {
+				// collect all conditions for a wildcard
+				r.inv.AddCondition(Wildcard, &objRef, &c)
+			} else {
+				for ref := range r.cfg.Owns {
+					if strings.Contains(c.Type, kptfilelibv1.GetConditionType(&ref)) {
+						r.inv.AddCondition(objRef.Name, &objRef, &c)
+					}
+				}
+			}
+		}
+	}
+
+	// filter the related resources per name in case no wildcard
+	// for wildcard we add all resources to the wildcard context
+	for _, o := range r.rl.GetObjects() {
+		if len(r.cfg.Owns) == 0 {
+			if o.GetAPIVersion() == r.cfg.For.ObjectRef.APIVersion && o.GetKind() == r.cfg.For.ObjectRef.Kind {
+				r.inv.AddForResource(o.GetName(), o)
+			} else {
+				r.inv.AddResource(Wildcard, &corev1.ObjectReference{
+					APIVersion: o.GetAPIVersion(),
+					Kind:       o.GetKind(),
+					Name:       o.GetName(),
+				}, o)
+			}
+		} else {
+			for objRef := range r.cfg.Owns {
+				if o.GetAPIVersion() == objRef.APIVersion && o.GetKind() == objRef.Kind {
+					r.inv.AddResource(o.GetName(), &corev1.ObjectReference{
+						APIVersion: o.GetAPIVersion(),
+						Kind:       o.GetKind(),
+						Name:       o.GetName(),
+					}, o)
+				}
+				if o.GetAPIVersion() == r.cfg.For.ObjectRef.APIVersion && o.GetKind() == r.cfg.For.ObjectRef.Kind {
+					r.inv.AddForResource(o.GetName(), o)
+				}
+			}
+		}
+	}
+}
+
+func (r *downstreamFnRuntime) update() {
+	kf := kptfilelibv1.NewMutator(r.rl.GetObjects()[0].String())
+	var err error
+	if _, err = kf.UnMarshal(); err != nil {
+		fn.Log("error unmarshal kptfile")
+		r.rl.AddResult(err, r.rl.GetObjects()[0])
+	}
+
+	for name, readyCtx := range r.inv.GetReadinessStatus() {
+		fn.Logf("name: %s condition: %s ready: %t\n", name, readyCtx.ForCondition.Type, readyCtx.Ready)
+		if readyCtx.Ready {
+			// generate the obj irrespective of current status
+			if r.cfg.For.GenerateFn != nil {
+				o, err := r.cfg.For.GenerateFn(readyCtx.Objs)
+				if err != nil {
+					r.rl.AddResult(err, o)
+				} else {
+					if readyCtx.ForObj != nil {
+						r.rl.SetObject(o)
+					} else {
+						r.rl.AddObject(o)
+					}
+					// set status to True
+					readyCtx.ForCondition.Status = kptv1.ConditionTrue
+					kf.SetConditions(readyCtx.ForCondition)
+				}
+			}
+		} else {
+			// if obj exists delete it
+			if readyCtx.ForObj != nil {
```

Comment:
If resource is not ready, then don't we need to update the condition with reason?

Reply:
yes I added this.